### DISCUSSION TITLE
Mikroelektronika Mini M4 STM32 issues

### DISCUSSION THREAD

**Discussion by ITstreet1** [stm32duino/Arduino_Core_STM32#2144] [Q&A]
Hi,

I have a board from the title. It uses STM32F415RG MCU with a pre-programmed HID bootloader.
https://www.mikroe.com/mini-stm32f4

While using a HID bootloader method to upload a sketch over the USB, it throws an error that the IDE can not detect the board. It does not show up on the COM port. Tried with the Reset button, but no luck.
What could be a problem? Maybe the onboard bootloader is not compatible with this STM32 support?

I can always program it with a UART adapter using Arduino IDE, but thought it should be better this way.

The board works like a charm using Necto Studio.

**Comment by fpistm** [comment]
The USB CDC have to be enabled to allow restart in HID bootloader.  You have to flash it the first time with USB enabled then it should be OK. Else probably it is not compatible.

**Comment by ITstreet1** [comment]
Tried all options for USB support. All I get is 

> Sketch uses 26228 bytes (2%) of program storage space. Maximum is 1048576 bytes.
> Global variables use 5036 bytes (3%) of dynamic memory, leaving 126036 bytes for local variables. Maximum is 131072 bytes.
> 
> +-----------------------------------------------------------------------+
> |         HID-Flash v2.2.1 - STM32 HID Bootloader Flash Tool            |
> |     (c)      2018 - Bruno Freitas       http://www.brunofreitas.com   |
> |     (c) 2018-2019 - Vassilis Serasidis  https://www.serasidis.gr      |
> |   Customized for STM32duino ecosystem   https://www.stm32duino.com    |
> +-----------------------------------------------------------------------+
> 
> > Trying to open the [COM9]...
> > Unable to open the [COM9]
> > Searching for [1209:BEBA] device...
> ##########
> Error - [1209:BEBA] device is not found :(> Searching for [COM9] ...
> > Finish
> the selected serial port > Finish
>  does not exist or your board is not connected
> 

And this no matter if I try to enter the bootloader with a button or not.

PS
This is just a Blink sketch

**Comment by fpistm** [comment]
What is COM9? 
You have to start the Serial at least in the Setup. Else how do you want reset in HID mode?

**Comment by ITstreet1** [comment]
There is nothing on COM9, nor is anything selected as a COM PORT option.

```
void setup() {
  Serial.begin(9600);
  pinMode(PC12, OUTPUT);
}
void loop() {
  digitalWrite(PC12, HIGH);
  delay(1000);
  digitalWrite(PC12, LOW);
  delay(1000);
}
```
I get the same message as above. For USB support tried both CDC options. Nothing. Do I have to enable USART or something else?

**Comment by fpistm** [comment]
Hi,
I convert the issue as a discussion as it seems more a usage issue or simply not supported.
To have the HID BL functional you have to enable the CDC USB. In the `Setup()`, `Serial.begin()` should be called. Then a new COM port should be available. And finally this COM port should be selected in the menu. In that case, the board can be reset in HID BL mode thanks the script. If it do not succeed, there is nothing we can do.
Now this board is not part of the core so I guess you used Generic F415RG target which used the HSI as default clock input.
Using HSI to feed the PLL is not recommended with USB anyway it should works.
It sees the board you used have a 16 MHz HSE so we could redefine the `SystemClock_Config`:
```C++
extern "C" void SystemClock_Config(void)
{
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};

  /** Configure the main internal regulator output voltage
  */
  __HAL_RCC_PWR_CLK_ENABLE();
  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);

  /** Initializes the RCC Oscillators according to the specified parameters
  * in the RCC_OscInitTypeDef structure.
  */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSE;
  RCC_OscInitStruct.HSEState = RCC_HSE_ON;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSE;
  RCC_OscInitStruct.PLL.PLLM = 8;
  RCC_OscInitStruct.PLL.PLLN = 168;
  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
  RCC_OscInitStruct.PLL.PLLQ = 7;
  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
    Error_Handler();
  }

  /** Initializes the CPU, AHB and APB buses clocks
  */
  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
                                |RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV4;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV2;

  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_5) != HAL_OK) {
    Error_Handler();
  }
}
```

It requires also to redefine the default `[HSE_VALUE](https://github.com/stm32duino/Arduino_Core_STM32/blob/61a1680ef96871a38672b5d9ea57d7a9df9b3141/system/STM32F4xx/stm32f4xx_hal_conf_default.h#L99-L101)` using the [build_opt.h](https://github.com/stm32duino/Arduino_Core_STM32/wiki/Customize-build-options-using-build_opt.h):
`-DHSE_VALUE=16000000U`

So first thing to have is the USB CDC functional. If not then you can't have the HID BL upload method operational except if you manually enter in HID mode.
One other possible problem is if the HID BL was modified by Mikroe, as we assume the flash offset is:
`flash_offset=0x4000`
